### DISCUSSION TITLE
How to make it work with external apps?

### DISCUSSION THREAD

**Discussion by zenny** [RHVoice/RHVoice#320] [Q&A]
Hi,

Compiled RHvoice in voidlinux manually (reproduced for future references) as follows.

```
sudo git clone --recursive https://github.com/RHVoice/RHVoice.git /opt/RHVoice
cd /opt/RHVoice
sudo scons
sudo scons install
sudo ldconfig
```

However, there are two issues:

1. The TTS voice has speckles. How to avoid them? Fyi, here are no speckles when used `mimic`. Even the voices in  https://rhvoice.su/voices/ plays without a problem in browser.
2. Is there a way to make it work with external applications like okular, foliate or firefox among others? `/usr/local/bin/RHVoice-test -p alan` works alright in command line except speckles.

Thanks.

**Comment by alex19EP** [comment]
> 1. The TTS voice has speckles.

what speckles is?

> 2. Is there a way to make it work with external applications like okular, foliate or firefox among others?

rhvoice can b used as Speech-dispatcher module. so if application uses
[Speech-dispatcher](https://freebsoft.org/speechd) it can use rhvoice.

**Reply by zenny** [reply]
> > 1. The TTS voice has speckles.
> 
> what speckles is?

jitters/crackling I meant. How to get rid of them?

> 
> > 1. Is there a way to make it work with external applications like okular, foliate or firefox among others?
> 
> rhvoice can b used as Speech-dispatcher module. so if application uses
> [Speech-dispatcher](https://freebsoft.org/speechd) it can use rhvoice.

I have installed speech-dispatcher and also make the symbolic link as in the README.md. 

Fyi, I tried in`foliate` with `/usr/local/bin/RHVoice-test' and it works (but with cracklings and jitters), but there is no way that `okular` picked up despite espeak-ng is already installed. The option field for speech-engine in `okular` preferences could not be selected.
```
$ spd-say --list-output-modules
OUTPUT MODULES
rhvoice
espeak-ng
```

**Reply by alex19EP** [reply]
> > > 1. The TTS voice has speckles.
> > 
> > 
> > what speckles is?
> 
> jitters/crackling I meant. How to get rid of them?

hmm they shouldn't be there. what is your sound system ALSA, pulce, etc?

also can you provide start of scons output?

> 
> > > 1. Is there a way to make it work with external applications like okular, foliate or firefox among others?
> > 
> > 
> > rhvoice can b used as Speech-dispatcher module. so if application uses
> > [Speech-dispatcher](https://freebsoft.org/speechd) it can use rhvoice.
> 
> I have installed speech-dispatcher and also make the symbolic link as in the README.md.
> 
> Fyi, I tried in`foliate` with `/usr/local/bin/RHVoice-test' and it works (but with cracklings and jitters),

can you test RHVoice with spd-say command?

> but there is no way that `okular`picked up despite espeak-ng is already installed. The option field for speech-engine in`okular` preferences could not be selected.
> 
> ```
> $ spd-say --list-output-modules
> OUTPUT MODULES
> rhvoice
> espeak-ng
> ```

then it is the problem of the okular program itself. they should add support for Speech-dispatcher.

**Reply by zenny** [reply]
> > > > 1. The TTS voice has speckles.
> > > 
> > > 
> > > what speckles is?
> > 
> > 
> > jitters/crackling I meant. How to get rid of them?
> 
> hmm they shouldn't be there. what is your sound system ALSA, pulce, etc?
> 
> also can you provide start of scons output?
> 
> > > > 1. Is there a way to make it work with external applications like okular, foliate or firefox among others?
> > > 
> > > 
> > > rhvoice can b used as Speech-dispatcher module. so if application uses
> > > [Speech-dispatcher](https://freebsoft.org/speechd) it can use rhvoice.
> > 
> > 
> > I have installed speech-dispatcher and also make the symbolic link as in the README.md.
> > Fyi, I tried in`foliate` with `/usr/local/bin/RHVoice-test' and it works (but with cracklings and jitters),
> 
> can you test RHVoice with spd-say command?

I confirm that `spd-say` output has no cracklings at all, only `RHVoice-test` output has.

Appended with `AudioOutputMethod "libao"` as suggested in https://askubuntu.com/questions/1027885/when-the-default-text-to-speech-program-spd-say-speaks-it-is-accompanied-by-no/1027888#1027888, but didn't help.

Appending `tsched=0` into `/etc/pulse/default.pa` as described in https://wiki.archlinux.org/index.php/PulseAudio/Troubleshooting#Glitches.2C_skips_or_crackling prevents starting `pulseaudio`. :(

I can post the outputs and commands if that could be helpful to overcome this problem.

> 
> > but there is no way that `okular`picked up despite espeak-ng is already installed. The option field for speech-engine in`okular` preferences could not be selected.
> > ```
> > $ spd-say --list-output-modules
> > OUTPUT MODULES
> > rhvoice
> > espeak-ng
> > ```
> 
> then it is the problem of the okular program itself. they should add support for Speech-dispatcher.

Thanks. `okular` reportedly uses espeak to access to `speech-dispatcher` according to https://www.ubuntubuzz.com/2018/12/text-to-speech-on-gnulinux-part-2-okular-to-speech.html, but this does not seem working anymore, fyi.

**Reply by zenny** [reply]
> > > > > 1. The TTS voice has speckles.
> > > > 
> > > > 
> > > > what speckles is?
> > > 
> > > 
> > > jitters/crackling I meant. How to get rid of them?
> > 
> > 
> > hmm they shouldn't be there. what is your sound system ALSA, pulce, etc?
> > also can you provide start of scons output?
> > > > > 1. Is there a way to make it work with external applications like okular, foliate or firefox among others?
> > > > 
> > > > 
> > > > rhvoice can b used as Speech-dispatcher module. so if application uses
> > > > [Speech-dispatcher](https://freebsoft.org/speechd) it can use rhvoice.
> > > 
> > > 
> > > I have installed speech-dispatcher and also make the symbolic link as in the README.md.
> > > Fyi, I tried in`foliate` with `/usr/local/bin/RHVoice-test' and it works (but with cracklings and jitters),
> > 
> > 
> > can you test RHVoice with spd-say command?
> 
> I confirm that `spd-say` output has no cracklings at all, only `RHVoice-test` output has.
> 
> Appended with `AudioOutputMethod "libao"` as suggested in https://askubuntu.com/questions/1027885/when-the-default-text-to-speech-program-spd-say-speaks-it-is-accompanied-by-no/1027888#1027888, but didn't help.
> 
> Appending `tsched=0` into `/etc/pulse/default.pa` as described in https://wiki.archlinux.org/index.php/PulseAudio/Troubleshooting#Glitches.2C_skips_or_crackling prevents starting `pulseaudio`. :(
> 
> I can post the outputs and commands if that could be helpful to overcome this problem.

The crackling problem was solved after creating local config with `spd-conf` and then, appended the ~/.config/speech-dispatcher/speechd.conf` with `AudioPulseMinLength 1764` (kudos to https://github.com/brailcom/speechd/issues/198#issuecomment-619605269). So crackling appears as a problem of `speech-dispatcher` to work with `pulseaudio`, not to do with RHVoice!


> 
> > > but there is no way that `okular`picked up despite espeak-ng is already installed. The option field for speech-engine in`okular` preferences could not be selected.
> > > ```
> > > $ spd-say --list-output-modules
> > > OUTPUT MODULES
> > > rhvoice
> > > espeak-ng
> > > ```
> > 
> > 
> > then it is the problem of the okular program itself. they should add support for Speech-dispatcher.
> 
> Thanks. `okular` reportedly uses espeak to access to `speech-dispatcher` according to https://www.ubuntubuzz.com/2018/12/text-to-speech-on-gnulinux-part-2-okular-to-speech.html, but this does not seem working anymore, fyi.

This appears to be a bottleneck with `okular` with No TTS engines are detected (https://github.com/void-linux/void-packages/issues/31608). So closing.
